### PR TITLE
Open ended capstone

### PR DESCRIPTION
**Pull request by AlanHorowitz**
Please review second_etl.py and comment.   The code demonstrates initial and incremental load of source data via psycopg2.
README.md has instructions to run the program and postgres dockerized.  My planned next steps are to add the created/updated date columns, upsert to mysql, and improved CLI.

### CODE REVIEW

File: open-ended-capstone/second_etl.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+import random
+import psycopg2
+import psycopg2.extras
+
+total_inserts = 0
+total_updates = 0
+
+class Column():
```
Comment (denysondata) [inline open-ended-capstone/second_etl.py]: redundant parenthesis

File: open-ended-capstone/second_etl.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    def get_column_names(self):
+
+        return [col.get_name() for col in self._columns] 
+
+    def get_update_column(self):
+
+        i = random.randint(0,len(self._update_columns)-1)   
+        return self._update_columns[i]
+
+PRODUCT_CREATE_SQL_PG = \
```
Comment (denysondata) [inline open-ended-capstone/second_etl.py]: use tripple quotes instead of backslashes

File: open-ended-capstone/second_etl.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import random
```
Comment (denysondata) [inline open-ended-capstone/second_etl.py]: code is not readable—variables, classes and functions are all mixed up.

no clearly visible entry point

Comment (denysondata) [inline open-ended-capstone/second_etl.py]: some variables (like total_inserts) are defined multiple times

Comment (denysondata) [inline open-ended-capstone/second_etl.py]: there are lines of code that do nothing

like.  `create_table_trg(trg_conn, None)`

File: open-ended-capstone/second_etl.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+" product_preferred_supplier_id INTEGER NULL DEFAULT NULL,"\
+" product_dimension_length FLOAT(11) NULL DEFAULT NULL,"\
+" product_dimension_width FLOAT(11) NULL DEFAULT NULL,"\
+" product_dimension_height FLOAT(11) NULL DEFAULT NULL,"\
+" product_introduced_date DATE NULL DEFAULT NULL,"\
+" product_discontinued BOOLEAN NULL DEFAULT FALSE,"\
+" product_no_longer_offered BOOLEAN NULL DEFAULT FALSE,"\
+" PRIMARY KEY (product_id));" 
+
+PRODUCT_TABLE = Table('product',
```
Comment (denysondata) [inline open-ended-capstone/second_etl.py]: not clear why you need a DDL statement AND a class to define your table

File: open-ended-capstone/second_etl.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    conn.commit()
+    cur.close()
+
+def create_table_trg(conn, table):
+    pass
+
+def extract_src_to_trg(source, target, table):
+    pass
+
+def incremental_load_src(src_conn, table, n_inserts, n_updates):
```
Comment (denysondata) [inline open-ended-capstone/second_etl.py]: not clear what types of objects are expected to be passed as variables. 

would recommend using types annotations

**Review by denysondata — CHANGES_REQUESTED** [review]
You could start with cleaning the code up and following some basic "narrative" in the code—for example the way it is discussed here https://dbader.org/blog/how-to-structure-python-programs


Then you could make sure that the code adheres to something like pep8 or black formatting style.

Then you could do variables and functions annotations https://docs.python.org/3/library/typing.html

Plus, make sure that only the code that you actually execute is present in your python files. 

After all of the above is done I could start looking into the code logic.

**Comment by AlanHorowitz** [comment]
The best way I see to clarify the narrative is to split the code into different  .py files by functional category. Sent from Mail for Windows 10 From: Denys StetsenkoSent: Sunday, March 14, 2021 11:36 AMTo: AlanHorowitz/SpringBoardCc: AlanHorowitz; AuthorSubject: Re: [AlanHorowitz/SpringBoard] Open ended capstone ***@***.*** requested changes on this pull request.You could start with cleaning the code up and following some basic "narrative" in the code—for example the way it is discussed here https://dbader.org/blog/how-to-structure-python-programsThen you could make sure that the code adheres to something like pep8 or black formatting style.Then you could do variables and functions annotations https://docs.python.org/3/library/typing.htmlPlus, make sure that only the code that you actually execute is present in your python files.After all of the above is done I could start looking into the code logic.In open-ended-capstone/second_etl.py:> @@ -0,0 +1,194 @@ +import random +import psycopg2 +import psycopg2.extras + +total_inserts = 0 +total_updates = 0 + +class Column(): redundant parenthesisIn open-ended-capstone/second_etl.py:> +        return self._columns + +    def get_primary_key(self): + +        return self._primary_key + +    def get_column_names(self): + +        return [col.get_name() for col in self._columns]  + +    def get_update_column(self): + +        i = random.randint(0,len(self._update_columns)-1)    +        return self._update_columns[i] + +PRODUCT_CREATE_SQL_PG = \ use tripple quotes instead of backslashesIn open-ended-capstone/second_etl.py:> @@ -0,0 +1,194 @@ +import random code is not readable—variables, classes and functions are all mixed up.no clearly visible entry pointIn open-ended-capstone/second_etl.py:> @@ -0,0 +1,194 @@ +import random some variables (like total_inserts) are defined multiple timesIn open-ended-capstone/second_etl.py:> @@ -0,0 +1,194 @@ +import random there are lines of code that do nothinglike. create_table_trg(trg_conn, None)In open-ended-capstone/second_etl.py:> +" CREATE TABLE IF NOT EXISTS Product ("\ +" product_id INTEGER NOT NULL,  "\ +" product_name VARCHAR(80) NOT NULL,"\ +" product_description VARCHAR(255) NULL DEFAULT NULL,"\ +" product_category VARCHAR(80) NULL DEFAULT NULL,"\ +" product_brand VARCHAR(80) NULL DEFAULT NULL,"\ +" product_preferred_supplier_id INTEGER NULL DEFAULT NULL,"\ +" product_dimension_length FLOAT(11) NULL DEFAULT NULL,"\ +" product_dimension_width FLOAT(11) NULL DEFAULT NULL,"\ +" product_dimension_height FLOAT(11) NULL DEFAULT NULL,"\ +" product_introduced_date DATE NULL DEFAULT NULL,"\ +" product_discontinued BOOLEAN NULL DEFAULT FALSE,"\ +" product_no_longer_offered BOOLEAN NULL DEFAULT FALSE,"\ +" PRIMARY KEY (product_id));"  + +PRODUCT_TABLE = Table('product',  not clear why you need a DDL statement AND a class to define your tableIn open-ended-capstone/second_etl.py:> +                          'BOOLEAN' : True } + +def create_table_src(conn, create_sql): + +    cur = conn.cursor() +    cur.execute(create_sql) +    conn.commit() +    cur.close() + +def create_table_trg(conn, table): +    pass + +def extract_src_to_trg(source, target, table): +    pass + +def incremental_load_src(src_conn, table, n_inserts, n_updates): not clear what types of objects are expected to be passed as variables.would recommend using types annotations—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

File: open-ended-capstone/RetailDW/etlutils.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    if n_updates > 0:
+
+        n_updates = min(n_updates, row_count)
+        update_keys = ",".join(
+            [str(i) for i in random.sample(range(1, next_key), n_updates)]
+        )
+        cur.execute(
+            f"SELECT {column_names} from {table.get_name()}"
+            f" WHERE {table.get_primary_key()} IN ({update_keys});"
+        )
```
Comment (denysthegitmenace) [inline open-ended-capstone/RetailDW/etlutils.py:72]: To discuss as part of point. 1 from the email.

**Comment by AlanHorowitz** [comment]
I think we are ready to close and merge this one, right?

**Review by denysondata — APPROVED** [review]

**Pull request merged**